File: app/src/main/java/com/galeria/android/FolderPickerActivity.java
package com.galeria.android;

import android.app.Activity;
import android.app.AlertDialog;
import android.os.Bundle;
import android.os.Environment;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class FolderPickerActivity extends Activity {
    private File rootDir;
    private File currentDir;
    private FolderAdapter adapter;
    private TextView pathView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        rootDir = Environment.getExternalStorageDirectory();
        File pictures = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
        currentDir = pictures.exists() ? pictures : rootDir;
        buildLayout();
        loadFolders();
    }

    private void buildLayout() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setBackgroundColor(Ui.bg(this));

        LinearLayout bar = new LinearLayout(this);
        bar.setGravity(Gravity.CENTER_VERTICAL);
        Ui.setPadding(bar, 12, 12, 12, 8);

        TextView back = Ui.title(this, "Voltar", 16);
        back.setGravity(Gravity.CENTER);
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });
        bar.addView(back, new LinearLayout.LayoutParams(Ui.dp(this, 76), Ui.dp(this, 44)));

        TextView title = Ui.title(this, "Selecionar pasta", 22);
        bar.addView(title, new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1));
        root.addView(bar);

        pathView = Ui.label(this, "");
        pathView.setGravity(Gravity.LEFT);
        Ui.setPadding(pathView, 18, 0, 18, 8);
        root.addView(pathView);

        ListView list = new ListView(this);
        list.setDivider(null);
        list.setCacheColorHint(android.graphics.Color.TRANSPARENT);
        list.setBackgroundColor(Ui.bg(this));
        adapter = new FolderAdapter();
        list.setAdapter(adapter);
        list.setOnItemClickListener((parent, view, position, id) -> {
            File item = adapter.getItem(position);
            if (item == null) {
                File parentDir = currentDir.getParentFile();
                if (parentDir != null && isInsideRoot(parentDir)) {
                    currentDir = parentDir;
                    loadFolders();
                }
            } else {
                currentDir = item;
                loadFolders();
            }
        });
        root.addView(list, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1));

        LinearLayout actions = new LinearLayout(this);
        actions.setGravity(Gravity.RIGHT | Gravity.CENTER_VERTICAL);
        Ui.setPadding(actions, 12, 10, 12, 14);

        TextView create = Ui.title(this, "Criar pasta", 16);
        create.setTextColor(Ui.accent(this));
        create.setGravity(Gravity.CENTER);
        create.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                askCreateFolder();
            }
        });
        actions.addView(create, new LinearLayout.LayoutParams(Ui.dp(this, 130), Ui.dp(this, 44)));

        TextView ok = Ui.title(this, "OK", 16);
        ok.setTextColor(Ui.accent(this));
        ok.setGravity(Gravity.CENTER);
        ok.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Ui.toast(FolderPickerActivity.this, "Pasta selecionada: " + currentDir.getName());
                finish();
            }
        });
        actions.addView(ok, new LinearLayout.LayoutParams(Ui.dp(this, 70), Ui.dp(this, 44)));
        root.addView(actions);

        setContentView(root);
    }

    private void loadFolders() {
        pathView.setText(currentDir.getAbsolutePath());
        ArrayList<File> folders = new ArrayList<>();
        if (!currentDir.equals(rootDir)) {
            folders.add(null);
        }
        File[] files = currentDir.listFiles();
        if (files != null) {
            Arrays.sort(files, new Comparator<File>() {
                @Override
                public int compare(File first, File second) {
                    return first.getName().compareToIgnoreCase(second.getName());
                }
            });
            for (File file : files) {
                if (file.isDirectory() && !file.isHidden()) {
                    folders.add(file);
                }
            }
        }
        adapter.submit(folders);
    }

    private boolean isInsideRoot(File dir) {
        return dir != null && dir.getAbsolutePath().startsWith(rootDir.getAbsolutePath());
    }

    private void askCreateFolder() {
        final EditText input = new EditText(this);
        input.setSingleLine(true);
        input.setHint("Nome da pasta");
        input.setTextColor(Ui.text(this));
        input.setHintTextColor(Ui.muted(this));
        new AlertDialog.Builder(this)
                .setTitle("Criar nova pasta")
                .setView(input)
                .setNegativeButton("Cancelar", null)
                .setPositiveButton("Criar", (dialog, which) -> createFolder(input.getText().toString()))
                .show();
    }

    private void createFolder(String name) {
        String clean = MediaActions.cleanFolderName(name);
        if (clean.isEmpty()) {
            Ui.toast(this, "Digite um nome valido.");
            return;
        }
        File target = new File(currentDir, clean);
        if (target.exists()) {
            Ui.toast(this, "A pasta ja existe.");
            return;
        }
        if (target.mkdirs()) {
            Ui.toast(this, "Pasta criada.");
            currentDir = target;
            loadFolders();
        } else {
            Ui.toast(this, "Nao foi possivel criar a pasta aqui.");
        }
    }

    private final class FolderAdapter extends BaseAdapter {
        private final ArrayList<File> folders = new ArrayList<>();

        void submit(List<File> nextFolders) {
            folders.clear();
            folders.addAll(nextFolders);
            notifyDataSetChanged();
        }

        @Override
        public int getCount() {
            return folders.size();
        }

        @Override
        public File getItem(int position) {
            return folders.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            LinearLayout row = new LinearLayout(FolderPickerActivity.this);
            row.setGravity(Gravity.CENTER_VERTICAL);
            Ui.setPadding(row, 18, 10, 18, 10);

            ImageView icon = new ImageView(FolderPickerActivity.this);
            icon.setImageResource(com.galeria.android.R.drawable.ic_folder);
            icon.setColorFilter(Ui.muted(FolderPickerActivity.this));
            row.addView(icon, new LinearLayout.LayoutParams(Ui.dp(FolderPickerActivity.this, 42), Ui.dp(FolderPickerActivity.this, 42)));

            TextView text = Ui.title(FolderPickerActivity.this, labelFor(getItem(position)), 18);
            text.setSingleLine(false);
            LinearLayout.LayoutParams textParams = new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1);
            textParams.leftMargin = Ui.dp(FolderPickerActivity.this, 14);
            row.addView(text, textParams);
            return row;
        }

        private String labelFor(File file) {
            if (file == null) {
                return "..";
            }
            File[] files = file.listFiles();
            int count = files == null ? 0 : files.length;
            return file.getName() + "\n" + count + (count == 1 ? " item" : " itens");
        }
    }
}

File: app/src/main/kotlin/com/galeria/android/FolderPickerActivity.kt
package com.galeria.android

import android.app.Activity
import android.app.AlertDialog
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ListView
import android.widget.TextView
import java.io.File

class FolderPickerActivity : Activity() {
    private lateinit var rootDir: File
    private lateinit var currentDir: File
    private lateinit var adapter: FolderAdapter
    private lateinit var pathView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        rootDir = Environment.getExternalStorageDirectory()
        val pictures = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
        currentDir = if (pictures.exists()) pictures else rootDir
        buildLayout()
        loadFolders()
    }

    private fun buildLayout() {
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(Ui.bg(this@FolderPickerActivity))
        }

        val bar = LinearLayout(this).apply {
            gravity = Gravity.CENTER_VERTICAL
            Ui.setPadding(this, 12, 12, 12, 8)
        }
        val back = Ui.title(this, "Voltar", 16).apply {
            gravity = Gravity.CENTER
            setOnClickListener { finish() }
        }
        bar.addView(back, LinearLayout.LayoutParams(Ui.dp(this, 76), Ui.dp(this, 44)))

        val title = Ui.title(this, "Selecionar pasta", 22)
        bar.addView(title, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))
        root.addView(bar)

        pathView = Ui.label(this, "").apply { gravity = Gravity.LEFT }
        Ui.setPadding(pathView, 18, 0, 18, 8)
        root.addView(pathView)

        val list = ListView(this).apply {
            divider = null
            cacheColorHint = android.graphics.Color.TRANSPARENT
            setBackgroundColor(Ui.bg(this@FolderPickerActivity))
        }
        adapter = FolderAdapter()
        list.adapter = adapter
        list.setOnItemClickListener { _, _, position, _ ->
            val item = adapter.getItem(position)
            if (item == null) {
                val parentDir = currentDir.parentFile
                if (parentDir != null && isInsideRoot(parentDir)) {
                    currentDir = parentDir
                    loadFolders()
                }
            } else {
                currentDir = item
                loadFolders()
            }
        }
        root.addView(list, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f))

        val actions = LinearLayout(this).apply {
            gravity = Gravity.RIGHT or Gravity.CENTER_VERTICAL
            Ui.setPadding(this, 12, 10, 12, 14)
        }
        val create = Ui.title(this, "Criar pasta", 16).apply {
            setTextColor(Ui.accent(this@FolderPickerActivity))
            gravity = Gravity.CENTER
            setOnClickListener { askCreateFolder() }
        }
        actions.addView(create, LinearLayout.LayoutParams(Ui.dp(this, 130), Ui.dp(this, 44)))

        val ok = Ui.title(this, "OK", 16).apply {
            setTextColor(Ui.accent(this@FolderPickerActivity))
            gravity = Gravity.CENTER
            setOnClickListener {
                Ui.toast(this@FolderPickerActivity, "Pasta selecionada: ${currentDir.name}")
                finish()
            }
        }
        actions.addView(ok, LinearLayout.LayoutParams(Ui.dp(this, 70), Ui.dp(this, 44)))
        root.addView(actions)
        setContentView(root)
    }

    private fun loadFolders() {
        pathView.text = currentDir.absolutePath
        val folders = ArrayList<File?>()
        if (currentDir != rootDir) {
            folders.add(null)
        }
        currentDir.listFiles()
            ?.sortedBy { it.name.lowercase() }
            ?.filter { it.isDirectory && !it.isHidden }
            ?.let(folders::addAll)
        adapter.submit(folders)
    }

    private fun isInsideRoot(dir: File?): Boolean =
        dir != null && dir.absolutePath.startsWith(rootDir.absolutePath)

    private fun askCreateFolder() {
        Ui.showTextInputDialog(
            this,
            "Criar nova pasta",
            "Nome da pasta",
            positiveText = "Criar"
        ) { name ->
            createFolder(name)
        }
    }

    private fun createFolder(name: String) {
        val clean = MediaActions.cleanFolderName(name)
        if (clean.isEmpty()) {
            Ui.toast(this, "Digite um nome válido.")
            return
        }
        val target = File(currentDir, clean)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !MediaActions.hasAllFilesAccess(this)) {
            Ui.showConfirmationDialog(
                this,
                "Permitir gerenciamento de arquivos",
                "Para criar pastas no celular, ative o acesso total a arquivos para a Galeria.",
                "Permitir"
            ) { MediaActions.requestAllFilesAccess(this) }
            return
        }
        if (target.exists()) {
            Ui.toast(this, "A pasta já existe.")
            return
        }
        if (MediaActions.createFolder(this, target)) {
            Ui.toast(this, "Pasta criada.")
            currentDir = target
            loadFolders()
        } else {
            Ui.toast(this, "Não foi possível criar a pasta aqui.")
        }
    }

    private inner class FolderAdapter : BaseAdapter() {
        private val folders = ArrayList<File?>()

        fun submit(nextFolders: List<File?>) {
            folders.clear()
            folders.addAll(nextFolders)
            notifyDataSetChanged()
        }

        override fun getCount(): Int = folders.size

        override fun getItem(position: Int): File? = folders[position]

        override fun getItemId(position: Int): Long = position.toLong()

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val row = LinearLayout(this@FolderPickerActivity).apply {
                gravity = Gravity.CENTER_VERTICAL
                Ui.setPadding(this, 18, 10, 18, 10)
            }

            val icon = ImageView(this@FolderPickerActivity).apply {
                setImageResource(R.drawable.ic_folder)
                setColorFilter(Ui.muted(this@FolderPickerActivity))
            }
            row.addView(
                icon,
                LinearLayout.LayoutParams(
                    Ui.dp(this@FolderPickerActivity, 42),
                    Ui.dp(this@FolderPickerActivity, 42)
                )
            )

            val text = Ui.title(this@FolderPickerActivity, labelFor(getItem(position)), 18).apply {
                isSingleLine = false
            }
            val textParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f).apply {
                leftMargin = Ui.dp(this@FolderPickerActivity, 14)
            }
            row.addView(text, textParams)
            return row
        }

        private fun labelFor(file: File?): String {
            if (file == null) return ".."
            val count = file.listFiles()?.size ?: 0
            return file.name + "\n" + count + if (count == 1) " item" else " itens"
        }
    }
}
